### PR TITLE
Gql parser 0.4

### PR DESCRIPTION
**Pull request by mathstuf**
---
Cc: @samuela

### CODE REVIEW

**Comment by mathstuf** [comment]
Also needs this PR now: https://github.com/graphql-rust/graphql-parser/pull/57

File: graphql_client_codegen/src/query.rs
Comment (mathstuf) [inline graphql_client_codegen/src/query.rs:663]: Needs https://github.com/graphql-rust/graphql-parser/pull/57

File: graphql_client_codegen/Cargo.toml
Comment (mathstuf) [inline graphql_client_codegen/Cargo.toml:12]: Needs released. Tracking: https://github.com/graphql-rust/graphql-parser/issues/50

File: graphql_client_codegen/src/query.rs
Comment (samuela) [reply graphql_client_codegen/src/query.rs:663]: If we start using `'static` things all over when will those values be freed? Does it happen when the corresponding `Value` gets freed? Using `'static` always strikes me as a little sketchy, but it may certainly be worth it here as long as it doesn't leak memory.

Comment (mathstuf) [reply graphql_client_codegen/src/query.rs:663]: It's `'static` because the *type* is not tied to a lifetime since it manages its own memory using `String` instead of borrowing from some other memory that lives…somewhere else. Basically, instead of `Cow<'a, str>` for any `'a` being allowed, we're saying `Cow<'static, str>` which is either a `&'static str` or an owned string. Both know when they need deallocated.

**Comment by tomhoule** [comment]
The extra lifetime and type parameters to do the same thing are a bit of a shame, but it's worth upgrading, thanks for the hard work! The PR looks good :+1: Let's wait for the graphql-parser PR to be merged and merge this.

File: graphql_client_codegen/src/query.rs
Comment (samuela) [reply graphql_client_codegen/src/query.rs:663]: My understanding is that an object having lifetime `'a` means that it lives at least as long as `'a`. So a lifetime of `'static` implies that the object must live for the lifetime of the program, no? Perhaps I'm missing something...

Comment (mathstuf) [reply graphql_client_codegen/src/query.rs:663]: For values, that's true. For *types*, a lifetime is the maximum time the type is valid for. The [`std::Error::source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) method requires a `'static` bound on its return in order to support downcasting. Since lifetimes don't exist at runtime, the only way you can validly downcast is if the type is not tied to any specific lifetime itself because there's no way to know that some lifetime-bound type is actually living long enough to support the cast.

**Comment by jmhain** [comment]
Is this PR close to being merged? There are a [couple](https://rustsec.org/advisories/RUSTSEC-2019-0036.html) [advisories](https://rustsec.org/advisories/RUSTSEC-2020-0036.html) for `failure` which `graphql_parser` 0.2 depends on.

**Comment by samuela** [comment]
https://github.com/graphql-rust/graphql-parser/pull/57 has been merged for a few months now, so this one should be ready to go but I don't know what the status is

**Comment by mathstuf** [comment]
Sorry for the delay; I've rebased and fix the compilation error I had locally.

**Comment by mathstuf** [comment]
What happened to CI here?

**Comment by jk-gan** [comment]
Hi, may I know what is the status of this PR?

**Comment by mathstuf** [comment]
Waiting on reviewers/maintainers AFAIK.

**Review by tomhoule — APPROVED** [review]
I'm not a fan of the extra generics over different string types, but since we depend on graphql parser, let's upgrade. Thanks for the PR!

**Pull request merged**
